libgdf/src/device_column.cu

```diff
@@ -0,0 +1,120 @@
+#include <iostream>
+#include "device_column.cuh"
+#include "gdf_type_dispatcher.cuh"
+
+// Anonymous namespace
+namespace{
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief Kernel to construct a TypedDeviceColumn<T> on the device.
+   * 
+   * @Param[in,out] p A pointer to a device allocation large enough to hold a 
+   * DeviceColumn*. When the kernel completes, this allocation will contain
+   * a pointer to the newly allocated TypedDeviceColumn.
+   * @Param[in] col The gdf_column that will be used to initialize the TypedDeviceColumn
+   */
+  /* ----------------------------------------------------------------------------*/
+  template <typename T>
+  __global__ 
+  void allocator_kernel(DeviceColumn ** p, gdf_column col){
+    if(0 == threadIdx.x + blockIdx.x * blockDim.x)
+      *p = static_cast<DeviceColumn*>(new TypedDeviceColumn<T>(col));
+  }
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief  Kernel to destroy a DeviceColumn constructed on the device.
+   * 
+   * @Param p Pointer to a device allocation that contains a pointer to the 
+   * device-constructed DeviceColumn that will be destroyed
+   */
+  /* ----------------------------------------------------------------------------*/
+  __global__ 
+  void deallocator_kernel(DeviceColumn ** p){
+    if(0 == threadIdx.x + blockIdx.x * blockDim.x)
+      delete *p;
+  }
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief  Functor that constructs on the device a TypedDeviceColumn<T> with 
+   * the type T determined by the gdf_dtype enum in a gdf_column.
+   */
+  /* ----------------------------------------------------------------------------*/
+  struct device_column_allocator{
+
+    /* --------------------------------------------------------------------------*/
+    /** 
+     * @brief  Templated operator() that constructs the TypedColumn<T> on the device.
+     * 
+     * This operator() is invoked by the gdf_type_dispatcher with the appropriate 
+     * T determined by the gdf_column's gdf_dtype enum.
+     * 
+     * @Param p 
+     * @Param col
+     */
+    /* ----------------------------------------------------------------------------*/
+    template<typename T>
+    __host__
+    DeviceColumn ** operator()(gdf_column const& col){
+      
+      DeviceColumn** p;
+
+      cudaMalloc(&p, sizeof(DeviceColumn*));
+      
+      auto status = cudaGetLastError();
+      if(cudaSuccess != status)
+        std::cerr << "Failed to allocate DeviceColumn pointer.\n";
+      
+      cudaEvent_t sync{};
+      cudaEventCreate(&sync);
+
+      // Invokes kernel to construct the TypedColumn<T> on the device
+      allocator_kernel<T><<<1,1>>>(p, col);
+      cudaEventRecord(sync);
+
+      // Use events to ensure kernel is complete and object is constructed before
+      // returning
+      cudaEventSynchronize(sync);
```

Comment:
This is effectively `cudaDeviceSynchronize()` since it is done on the null stream. Perhaps add an optional parameter (which defaults to `0`) to the constructor of this class to store a stream to use in the `operator()`. That way it can be optimized in cases where a stream is available or the user wants to create one.

Reply:
Good idea. I've been mainly avoiding streams until we figured out some kind of standard way we want to do streams, but I think the change you suggest is a good one no matter what we come up with.